dask_ml/model_selection/_incremental.py

```diff
@@ -347,3 +394,548 @@ def fit(
+        # TODO: More evidence to move away from BaseSearchCV
+        # self.best_score_ = self.history_results_[best_index]
+        self.n_splits_ = 1
+        self.multimetric_ = False  # TODO: is this always true?
+        raise gen.Return(self)
+
+    @property
+    def best_score_(self):
+        check_is_fitted(self, "best_index_")
+        return self.history_results_[self.best_index_]["score"]
+
+    def fit(self, X, y, **fit_params):
+        """Find the best parameters for a particular model.
+
+        Parameters
+        ----------
+        X, y : array-like
+        **fit_params
+            Additional partial fit keyword arguments for the estimator.
+        """
+        return default_client().sync(self._fit, X, y, **fit_params)
+
+
+class RandomizedIncrementalSearch(BaseIncrementalSearch):
+    """Incremental optimization with randomly sampled hyperparameters.
+
+    Incrementally hyper-parameter optimization creates a batch of models
+    and repeatedly call `partial_fit` with batches of data. The performance
+    of each model is monitored over time.
+
+    The set of hyper-parameters considered are randomly sampled from the
+    provided `param_distribution`.
+
+    Models are prioritized according to three parameters: `tol`, `patience`
+    and `max_iter`. Models are repeatedly trained until their score on the
+    test set stops improving. The last `patience` (10 by default) partial
+    fit iterations are considered. If the improvement is small (less than)
+    `tol`, training on that model is halted. If `max_iter` iterations is
+    reached training is halted, regardless of whether the model is still
+    improving.
+
+    Parameters
+    ----------
+    estimator : estimator object.
+        A object of that type is instantiated for each grid point.
+        This is assumed to implement the scikit-learn estimator interface.
+        Either estimator needs to provide a `score`` function,
+        or ``scoring`` must be passed. The estimator must implement
+        ``partial_fit``.
+
+    param_distributions : dict
+        Dictionary with parameters names (string) as keys and distributions
+        or lists of parameters to try. Distributions must provide a ``rvs``
+        method for sampling (such as those from scipy.stats.distributions).
+        If a list is given, it is sampled uniformly.
+
+    n_iter : int, default=10
+        Number of parameter settings that are sampled. n_iter trades
+        off runtime vs quality of the solution.
+
+    max_iter : int, default 100
+        Maximum number of partial fit iterations per model. This is passed
+
+    patience : int, default 10
+        The number of previous models scores to consider when determining
+        whether the model has converged. The most recent score must be at
+        at most `tol` better than the all of the previous `patience` scores.
+        Increasing `patience` will tend to provide better models, at the cost
+        of increased computation.
```

Comment:
Maybe `patience` should be in terms of epochs/passes through data, not number of `partial_fit` calls (and a float instead of an int). It helps remove an implicit dependence on the chunk size. That's what other similar classes use – e.g., PyTorch's [`ReduceLROnPlateau`][1] defaults to `patience` being 10 epochs.

This would require having more metadata to record the number of examples the estimator has seen, which will involve changes to `partial_fit`.

[1]:https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.ReduceLROnPlateau

Reply:
Sometimes we train on too much data and would like to stop early.

I agree that losing a dependence on chunk size would be useful.  I don't think that we can go with full dataset size though.  Number of records?  Amount of time?  Amount of bytes?